### PR TITLE
CURATOR-358 - Fixed race condition with getLeader()

### PR DESCRIPTION
**Pull request by cammckenzie**
-If leadership changes between the getParticipantNodes() call and the getLeader() internal call the NoNodeException is now handled and the next child in the list is evaluated.

Another option would be to just return the default empty Participant object and not iterate over the whole list of participants.

### CODE REVIEW

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java [hunk 1]
```diff
@@ -341,11 +342,41 @@ public Participant getLeader() throws Exception
...
         if ( participantNodes.size() > 0 )
         {
-            return participantForPath(client, participantNodes.iterator().next(), true);
+            Iterator<String> iter = participantNodes.iterator();
+            while ( iter.hasNext() )
+            {
+                
+                try
+                {
+                    result = participantForPath(client, iter.next(), true);
```
Comment (dragonsinth) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: can probably just return here right?

Comment (Randgalt) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: Why not have `participantForPath()` handle the exception and return `null`?

Comment (cammckenzie) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: It's called from other places too, so I'll need to refactor those as well. Probably a better approach though.

Comment (dragonsinth) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: Hehe, I wasn't worried about efficiency, it was a code clarity thought.

Comment (cammckenzie) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: Any other comments?

**Comment by Randgalt** [comment]
LGTM

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java (hunk 1 above)
Comment (dragonsinth) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: Nah, looks good.  What I was talking about totally depends on which style you find more readable.  I'm not from the school of single-exit; to me multiple return points is easier to follow than mentally tracking variable state.  I would probably have written this:

```java
static Participant getLeader(CuratorFramework client, Collection<String> participantNodes) throws Exception {
    for (String node : participantNodes) {
        Participant result = participantForPath(client, node, true);
        if ( result != null )  {
            return result;
        }
    }
    return new Participant();
}
```

But what you got works just as well.

Comment (cammckenzie) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java]: Yep, fair call, it's probably more Java like to just return. I've probably been writing too much C code lately!

**Pull request merged**